### DISCUSSION TITLE
Touch graphics tablet as keyboard replacement

### DISCUSSION THREAD

**Discussion by schrmh** [iberianpig/fusuma#312] [General]
Hi,
I do have a Wacom CTH-670, which is a big multi-touch tablet. Works with fusuma.
I kinda want to be able to press on specific points on the keyboard to send specific keys. From the description of fusumu-sendkey, it doesn't look like that is possible (yet).
Any idea how I could achieve that functionality?

The goal for that is to get a functionality to replace a keyboard. The area of the tablet does fit the layout for all the letter keys.

**Comment by iberianpig** [comment]
I'm interested in learning more about your specific use case. 
I think that quite challenging to input on a touch device as if it were a keyboard.

btw, fusuma-plugin-sendkey is a plugin that allows for input as a virtual keyboard and can be used as an action within fusuma.

**Reply by schrmh** [reply]
The use case is that a full keyboard is not necessarily needed anymore for general computer activity.
Graphics tablet as keyboard replacement (at least partially).
I often use my computer while lying in bed and I would like to only have the graphics tablet lying on me and not graphics tablet and keyboard (that's a bit uncomfortable) and the space next to me is already reserved for a real mouse (could at maximum add a numpad or small touchpad I don't own there).

**Reply by iberianpig** [reply]
What kind of operations do you envision for using a graphics tablet as a keyboard replacement?

It might be difficult to use a graphics tablet for typing as a keyboard with the fusuma's use of the `libinput debug-events` command, which does not capture the absolute position of touches.

Developing a fusuma plugin that captures touch positions from other resource and returns different statuses based on the location might be a good solution.

**Reply by schrmh** [reply]
I would just see how far I can get without using an physical keyboard while chatting, browsing, writing comments, etc.
I don't have a problem with typing on a flat surface and I write without looking on my keyboard.

However, position can also be useful for less obscure use cases.
I actually found a bit of useful things I will look into in the near future I think.
https://github.com/Wacom-Developer/wacom-device-kit-linux/tree/master has examples that work with my tablet, however not explicitly for touch point positions (and when trying with xsetwacom I also did not get anything useful regarding that).
And https://github.com/whot/mtview shows where I touched which means position can somehow be received even for touch. (Interestingly, my tablet seems to support over 10 touch points).

**Reply by iberianpig** [reply]
Using evdev might be beneficial as it operates at a lower layer than libinput and should be able to provide touchpoint data.

**Reply by schrmh** [reply]
Yeah, it looks like that mtview makes use of evdev when detecting the touch positions for my tablet.

**Reply by schrmh** [reply]
Looks like this is about the minimum for the touch position detection
```c
#include <stdio.h>
#include <fcntl.h>
#include <linux/input.h> //or <mtdev.h>, work both
#include <unistd.h>

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s /dev/input/eventX\n", argv[0]);
        return 1;
    }

    char *input_device = argv[1];
    int fd = open(input_device, O_RDONLY);
    struct input_event ev;

    if (fd == -1) {
        perror("Error opening input device");
        return 1;
    }

    int touch_id = 0;

    while (1) {
        if (read(fd, &ev, sizeof(struct input_event)) == sizeof(struct input_event)) {
            if (ev.type == EV_ABS) {
                if (ev.code == ABS_MT_SLOT) {
                    touch_id = ev.value;
                }
                if ((ev.code == ABS_MT_POSITION_X || ev.code == ABS_MT_POSITION_Y)) {
                    printf("%d: %c: %d\n", touch_id, (ev.code == ABS_MT_POSITION_X) ? 'x' : 'y', ev.value);
                }
            }
        }
    }

    close(fd);
    return 0;
}
```
(btw. ID goes from 0 up to 15 for me so maybe the 16FG within the displayed input device name "Wacom Bamboo 16FG 6x8" does stand for "16 fingers"?)

**Reply by iberianpig** [reply]
I confirmed that coordinates on the touchpad have been taken with your C program.
You can get 16 touchpoints, can't you? 
BTW, I'm curious, how did you test it? 😄

**Reply by schrmh** [reply]
Ah I basically just put on my whole hands with angled fingers to see how high it will go. Previously when I tried mtview which doesn't display coordinates in text but shows colors, I used my nose as a eleventh touch point. And when it actually looked legit I was curious how high it will go.
It really is uncommon — well, graphics tablets with finger touch support aren't that common anyways I think? — and the only thing I saw so far that might have more touch points are big touch tables intended for use by multiple people...

(Btw. 6x8 might by a category by wacom for rough active area. However when measuring between the lines indicating the touch area, I get 21.8x14cm (outer lines;  -2 mm for nearer lines) instead of 15.24x20.32. And even when numbers are rounded, it wouldn't be consistent. However I also saw a site claiming that this thing has ~20.8 cm in dimensions somewhere...)
